src/kkt.jl

```diff
@@ -0,0 +1,66 @@
+using QuadraticModels, RipQP, SparseArrays
+
+export kkt_checker
+
+"""
+    kkt_checker(nlp, sol; kwargs...)
+
+Given an NLPModels `nlp` and a vector `sol`, it returns the KKT residual of an optimization problem as a tuple (primal, dual).
+In particular, it uses `ripqp` to solve the following quadratic optimization problem with linear constraints
+```
+min_{d} ∇f(sol)ᵀd +  ½∥d∥²
+        lvar ≤ sol + d ≤ uvar
+        lcon ≤ c(sol) + ∇c(sol)d ≤ ucon
+```
+Keyword arguments are passed to `RipQP`.
+"""
+function kkt_checker(
+  nlp::AbstractNLPModel{T, S},
+  sol;
+  kwargs...,
+) where {T, S}
+  nvar = nlp.meta.nvar
+  g = grad(nlp, sol)
+  Hrows, Hcols = collect(1:nvar), collect(1:nvar)
+  Hvals = ones(T, nlp.meta.nvar)
+  kkt_nlp = if nlp.meta.ncon > 0
+    c = cons(nlp, sol)
+    feas_res = vcat(
+      max.(nlp.meta.lcon - c, c - nlp.meta.ucon, 0),
+      max.(nlp.meta.lvar - sol, sol - nlp.meta.uvar, 0)
+    )
+    Arows, Acols = jac_structure(nlp)
+    Avals = jac_coord(nlp, sol)
+    QuadraticModel(
+      g,
+      Hrows,
+      Hcols,
+      Hvals,
+      Arows = Arows,
+      Acols = Acols,
+      Avals = Avals,
+      lcon = nlp.meta.lcon .- c,
+      ucon = nlp.meta.ucon .- c,
+      lvar = nlp.meta.lvar .- sol,
+      uvar = nlp.meta.uvar .- sol,
+      x0 = fill!(S(undef, nlp.meta.nvar), zero(T)),
+    )
+  else
+    feas_res = max.(nlp.meta.lvar - sol, sol - nlp.meta.uvar, 0)
```

Comment:
This line could be factored out.

Reply:
Thanks for the review. I moved the bound feasibility before the if-else condition to avoid its repetition.